lib/mayaUsd/ufe/UsdContextOps.cpp

```diff
@@ -173,6 +192,137 @@ struct WaitCursor
     ~WaitCursor() { MGlobal::executeCommand("waitCursor -state 0"); }
 };
 
+//! \brief This check has a 3 seconds slowdown to load all Sdr nodes in the registry. We do it in
+///        advance in order to not have this 3 seconds delay when the "Assign New Material" submenu
+///        is built for the first time.
+bool _hasArnoldShaders()
+{
+    auto findArnold = []() {
+        const auto& sdrRegistry = PXR_NS::SdrRegistry::GetInstance();
+        auto        sourceTypes = sdrRegistry.GetAllNodeSourceTypes();
+        return std::find(sourceTypes.cbegin(), sourceTypes.cend(), TfToken("arnold"))
+            != sourceTypes.cend();
+    };
+    static const bool kHasArnoldShaders = findArnold();
+    return kHasArnoldShaders;
+};
+
+struct MxShaderMenuEntry
+{
+    MxShaderMenuEntry(const std::string& label, const std::string& identifier)
+        : _label(label)
+        , _identifier(identifier)
+    {
+    }
+    std::string        _label;
+    const std::string& _identifier;
+};
+typedef std::vector<MxShaderMenuEntry> MxShaderMenuEntryVec;
+
+const MxShaderMenuEntryVec& getMaterialXSurfaceShaders()
+{
+    static MxShaderMenuEntryVec mxSurfaceShaders;
+    if (mxSurfaceShaders.empty()) {
+        auto& sdrRegistry = PXR_NS::SdrRegistry::GetInstance();
+        // Here is a list of nodes we know work fine as starting materials for the contextual menu.
+        // We might add discovery code later, but this discovery code will have the difficult task
+        // of filtering out:
+        //   - utility nodes like ND_add_surfaceshader
+        //   - basic building blocks like ND_thin_surface
+        //   - shaders that exist only as pure definitions like ND_disney_bsdf_2015_surface
+        const std::vector<std::string> vettedSurfaces = { "ND_standard_surface_surfaceshader",
+                                                          "ND_gltf_pbr_surfaceshader",
+                                                          "ND_UsdPreviewSurface_surfaceshader" };
+        for (auto&& identifier : vettedSurfaces) {
+            auto shaderDef = sdrRegistry.GetShaderNodeByIdentifier(TfToken(identifier));
+            if (!shaderDef) {
+                continue;
+            }
+            std::string label = UsdMayaUtil::prettifyName(shaderDef->GetFamily().GetString());
+            mxSurfaceShaders.emplace_back(label, shaderDef->GetIdentifier().GetString());
+        }
+    }
+    return mxSurfaceShaders;
+}
+
+//! \brief Create an Item and select it:
+class BaseCreateAndSelectUndoableCommand : public Ufe::UndoableCommand
+{
+public:
+    BaseCreateAndSelectUndoableCommand(const Ufe::UndoableCommand::Ptr& creationCmd)
+        : _creationCmd(creationCmd)
+    {
+    }
+
+    void redo() override
+    {
+        if (_selectionCmd) {
+            _creationCmd->redo();
```

Comment:
If creation fails, it is expected that it will have cleaned-up correctly and not left the scene in an invalid state. It should also report what happened in an error message. This means that the composite only has to check before trying to continue (as it does) and not select something that does not exist.

We still end up with a no-op event on the undo stack, that should be no-up as well when undone. Which is the case with this code.